library/Icinga/Web/RememberMe.php

```diff
@@ -0,0 +1,206 @@
+        $rememberMe->rsa = $rsa;
+
+        return $rememberMe;
+    }
+
+    /**
+     * Set the values for 'remember-me' cookie
+     *
+     * @return Cookie
+     */
+    public function getCookie()
+    {
+        $value[] = $this->rsa->encryptToBase64($this->username);
+        $value[] = $this->encryptedPassword;
+        $value[] = base64_encode($this->rsa->getPublicKey());
+
+        return (new Cookie(static::COOKIE))
+            ->setExpire(time() + 60 * 60 * 24 * 30)
+            ->setHttpOnly(true)
+            ->setValue(implode('|', $value));
+    }
+
+    /**
+     * Unset the values for 'remember-me' cookie
+     *
+     * @return Cookie
+     */
+    public static function unsetCookie()
+    {
+        unset($_COOKIE[static::COOKIE]);
+
+        return (new Cookie(static::COOKIE))
+            ->setHttpOnly(true)
+            ->forgetMe();
+    }
+
+    /**
+     * Authenticate the given username and password
+     *
+     * @return bool     True if authentication succeed, false if not
+     *
+     * @throws \Icinga\Exception\AuthenticationException
+     */
+
+    public function authenticate()
+    {
+        $password = $this->rsa->decryptFromBase64($this->encryptedPassword);
+        $auth = Auth::getInstance();
+        $authChain = $auth->getAuthChain();
+        $authChain->setSkipExternalBackends(true);
+        $user = new User($this->username);
+        if (! $user->hasDomain()) {
+            $user->setDomain(Config::app()->get('authentication', 'default_domain'));
+        }
+        $authenticated = $authChain->authenticate($user, $password);
+        if ($authenticated) {
+            $auth->setAuthenticated($user);
+        }
+
+        return $authenticated;
+    }
+
+    /**
+     * Database insert for the given private and public key
+     *
+     * Remove old entry for given user if exists
+     * Save new keys in database
+     */
+    public function persist()
+    {
+        $this->remove();
+        $this->getDb()->insert('rememberme', [
+            'username' => $this->username,
+            'private_key' => $this->rsa->getPrivateKey(),
+            'public_key' => $this->rsa->getPublicKey(),
+            'ctime' => new Expression('NOW()'),
+            'mtime' => new Expression('NOW()')
+        ]);
+    }
+
+    /**
+     * Delete database entry if user logout or new keys are created for the same user
+     *
+     * by logout, this class do not have the login data for user, so username must
+     * be given as parameter to delete the user information from database.
+     *
+     * @param null|$username
+     */
+    public function remove($username = null)
+    {
+        $this->getDb()->delete('rememberme', [
+            'username = ?' => $this->username ?: $username
+        ]);
+    }
+
+    /**
+     * Renew the cookie
+     *
+     * Delete old database entry and insert new data for the given user
+     *
+     * @return Cookie
+     */
+    public function renewCookie()
```

Comment:
This should be `renew()` as stated above. Maybe we should also not call `remove()` and `persist()` automatically here because that is really to much side effect for the function. And since `persist()` calls `remove()` anyway, renewing a remember cookie is as simple and expressive as

```php
$rememberMe
    ->renew()
    ->persist()
    ->getCookie();
```

What do you think?